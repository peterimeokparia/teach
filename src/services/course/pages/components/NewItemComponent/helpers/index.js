import { navigate } from '@reach/router';
import { v4 as uuidv4 } from 'uuid';
import { formTypes } from 'services/course/pages/FormBuilder/helpers';

    const generateUuid = () => {
        const uuid = uuidv4();

        return uuid;
    };

export const forms = ( lesson, typeOfForm, formProps ) => {
    buildFormBuilderRoute( lesson, typeOfForm, formProps  );    
};

function buildFormBuilderRoute( lesson, typeOfForm, formProps ){
    
    let { operatorBusinessName, currentUser } = formProps;

    let props = {
        operatorBusinessName, currentUser, uuid: generateUuid(), lesson,
        userId: currentUser?._id, courseId: lesson?.courseId, lessonId: lesson?._id
    };

    switch ( typeOfForm ) {
        case formTypes.quizzwithpoints:   
            handleQuizz(props);
            break;
        case formTypes.homework: 
            handleHomeWork(props);
            break;
        case formTypes.furtherstudy: 
            handleFurtherStudy(props);
            break;    
        case formTypes.lessoninsights: 
            handleLessonInsights(props);
            break;    
        default:
            break;     
    };
}

function handleQuizz(props){
    let { 
     operatorBusinessName, uuid, userId, courseId, lessonId, lesson
    } = props;

    let formName = getFormName(lesson,formTypes?.quizzwithpoints,uuid);
    let formType = formTypes?.quizzwithpoints;
    let navProps = { operatorBusinessName, formType, formName, courseId, lessonId, userId };

    handleFormNavigation(navProps);
}

function handleHomeWork(props){
    let { 
        operatorBusinessName, uuid, userId, courseId, lessonId, lesson
    } = props;

    let formName = getFormName(lesson, formTypes?.homework, uuid);
    let formType = formTypes?.homework; 
    let navProps = { operatorBusinessName, formType, formName, courseId, lessonId, userId };

    handleFormNavigation(navProps);
}

function handleFurtherStudy(props){
    let { 
        operatorBusinessName, uuid, userId, courseId, lessonId, lesson 
    } = props;

    let formName = getFormName(lesson, formTypes?.furtherstudy, uuid);
    let formType = formTypes.furtherstudy; 
    let navProps = { operatorBusinessName, formType, formName, courseId, lessonId, userId };

    handleFormNavigation(navProps);
}

function handleLessonInsights(props){
    let { 
     operatorBusinessName, uuid, userId, courseId, lessonId, lesson
    } = props;

    let formName = getFormName(lesson, formTypes?.lessoninsights, uuid);
    let formType = formTypes?.lessoninsights;
    let navProps = { operatorBusinessName, formType, formName, courseId, lessonId, userId };

    handleFormNavigation(navProps);
}

function getFormName(currentLesson, type, formUniqueId){
    return `${currentLesson?.title}-${type}_${formUniqueId}`
}

function handleFormNavigation(navProps){
    let {
        operatorBusinessName, formType, formName, courseId, lessonId, userId
    } = navProps;

    navigate(`/${operatorBusinessName}/${formType}/formname/${formName}/courseid/${courseId}/lessonid/${lessonId}/user/${userId}`);
}