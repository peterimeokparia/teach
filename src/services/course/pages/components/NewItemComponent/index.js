import { useState, useRef, useEffect } from 'react';
import { Validations } from 'services/course/helpers/Validations';
import './style.css';

const NewItemComponent = ({
    resetItemError, 
    saveItemInProgress, 
    item,
    items,
    error,
    className,
    onSubmit,
    deleteItemAction,
    setField,
    fieldValue,
    placeholder,
    initialValue,
    inputType,
    editItem,
    forms,
    children 
}) => {
    const [ editing, setEditing ] = useState(editItem);
    const inputRef = useRef();

    const reset = () => {
        setField(initialValue);
        setEditing(false);
        resetItemError();
    };

    const commitEdit = (e) => {
        e.preventDefault();

        if ( Validations.duplicateCheck( fieldValue,  items, "item fieldValue", "fieldValue" ) ) { // check
            return;
        }

        onSubmit( fieldValue )
        .then(reset)
        .catch( error => {
            setEditing(false);
            setEditing(true);
        });
    };

    const setFieldValue = () => {
        setField(initialValue);
    };

    const beginEditing = () => {
        setFieldValue();
        setEditing(true);
    };

    const performDelete = () => {
        deleteItemAction(item);
    };

    useEffect (() => {
        if ( editing ) {
            inputRef.current.focus();
        }
    }, [ editing ]); 

    if ( saveItemInProgress ){
        return <div>Save in progress, please wait.</div>;
    };

return editing 
        ? (
            <>
                <form
                    className= {`${className || ''} editing ${ error ? 'error' : '' }`}
                    onSubmit={e => commitEdit( e )}            
                >    
                    <input 
                        ref={ inputRef }
                        type={ inputType }
                        value={ fieldValue }
                        onChange={ e => setField( e.target.value) }
                        onBlur={ reset }
                        placeholder={ placeholder }
                    />
                </form>
                {error && <div>{error.message}</div>}
             </>
            ) 
        : ( 
                children( beginEditing, performDelete, forms )
        );                
};


export default NewItemComponent;