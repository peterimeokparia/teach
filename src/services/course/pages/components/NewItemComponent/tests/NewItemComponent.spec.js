import React from 'react';
import NewItemComponent from '..';
import renderer from 'react-test-renderer';
import thunk from "redux-thunk"; 
import * as reactRedux from 'react-redux';
import { Provider, useSelector } from "react-redux";
import { applyMiddleware } from 'redux'; 
import configureStore from 'redux-mock-store';
jest.mock('../../../../api');

const mockStore = configureStore([]);
  
let operatorBusinessName= "teach";
let selectedUserId= "PERSON010010100000047";

describe('NewItemComponent', () =>  {  

 let  store = mockStore({
      operators: { operators: {_id: "5fcb0e19fd5e0117dc09dcfa", operatorBusinessName} },
      notifications: { pushNotificationSubscribers : {_id:"5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" } },
      users: { users: {_id: "PERSON5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      courses: { courses: {_id: "COURSE5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      grades: { grades: {_id: "COURSE5fcb0e19fd5e0117dc09dcfa", operatorBusinessName, operatorId: "5fcb0e19fd5e0117dc09dcfa" }},
      classrooms: { classrooms: {displaySideBarDropDown: false }},
      sessions: { sessions: {_id: "SESSIONS5fcb0e19fd5e0117dc09dcfa", numberOfSessions: 9, totalNumberOfSessions: 1, status: true } },
      lessons: { lessons: { selectedLessonFromLessonPlanDropDown: "Test Lesson" } },
      courses: { courses: { selectedCourseFromLessonPlanCourseDropDown: "Test Course" }  },
      meetings: { meetings: {_id: "MEETING5fcb0e19fd5e0117dc09dcfa", numberOfSessions: 9, totalNumberOfSessions: 1, status: true }  },
  }, applyMiddleware(thunk));

  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
 
  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  it('renders as expected', async () => {

    const dispatch = jest.fn();
    useDispatchMock.mockReturnValue(dispatch)

    const tree = renderer.create(
      <Provider store={store}>
        <NewItemComponent
          editItem={true} 
        >
        {( editItem, deleteItem ) => <div></div>}
        </NewItemComponent> 
    </Provider>
    );
      expect(tree.toJSON()).toMatchSnapshot();
  })
});

