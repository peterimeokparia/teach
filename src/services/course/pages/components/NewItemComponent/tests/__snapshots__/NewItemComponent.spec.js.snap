// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItemComponent renders as expected 1`] = `
<form
  className=" editing "
  onSubmit={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
  />
</form>
`;
